File: tools/llm-bot/src/main/java/org/finos/springbot/tool/llm/ChatGPTLLMService.java
package org.finos.springbot.tool.llm;

import org.springframework.ai.chat.model.ChatModel;
import org.springframework.ai.chat.model.ChatResponse;
import org.springframework.ai.chat.prompt.Prompt;
import org.springframework.ai.openai.OpenAiChatOptions;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class ChatGPTLLMService implements LLMService {

    @Autowired
    ChatModel chatModel;

    @Override
    public String getResponse(String request) {
        ChatResponse response = chatModel.call(
                new Prompt(
                        request,
                        OpenAiChatOptions.builder()
                                .model("gpt-4o")
                                .temperature(0.4)
                                .build()));
        return response.toString();
    }
}

File: tools/llm-bot/src/main/kotlin/org/finos/springbot/tool/llm/ChatGPTLLMService.kt
package org.finos.springbot.tool.llm

import ai.koog.agents.core.agent.AIAgent
import ai.koog.prompt.executor.clients.openai.OpenAIModels
import ai.koog.prompt.executor.llms.all.simpleOpenAIExecutor
import kotlinx.coroutines.runBlocking
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
class ChatGPTLLMService() : AbstractLLMService() {

    @Value("\${openai.api-key:bob}") override lateinit var apiKey: String

    var agent: AIAgent<String, String>? = null

    override fun afterPropertiesSet() {
        super.afterPropertiesSet()
        agent =
                AIAgent(
                        executor = simpleOpenAIExecutor(apiKey),
                        systemPrompt =
                                "You are a helpful assistant. Answer user questions concisely.",
                        llmModel = OpenAIModels.Chat.GPT4o
                )
    }

    override fun getResponse(request: String): String = runBlocking {
        val result = agent?.run(request)
        result?.toString() ?: "No response"
        // "No response"
    }
}
